### PR TITLE
Problem: Disorganised wallet feature (fix #183)

### LINKED ISSUE #183: Problem: Disorganised wallet features
Few things,

1. wallets functions are coupled with DefiWalletCoreActor -> Decouple them, and turn wallet as wallet component, so that wallet can attach to any actor instead of only DefiWalletCoreActor.
2. Most of wallets are not incomplete ->  bind more wallet functions from play-cpp-sdk to unreal.

### PR DESCRIPTION
**Pull request by damoncro**
- Add WalletComponent

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-com/play-unreal-plugin/blob/main/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest main? 
- [ ] Have you checked your code compiles in Unreal Engine 4 and 5?
- [ ] Have you checked your basic code formatting and style is fine? ([using this Linter plugin](https://ue4-style-guide.readthedocs.io/en/latest/gettingstarted.html))
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the plugin version number and documented your changes in the [CHANGELOG.md](https://github.com/crypto-com/play-unreal-plugin/blob/main/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-com/play-unreal-plugin/blob/main/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-com/play-unreal-plugin/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Comment by leejw51crypto** [comment]
UWallet is redundant, 
how about directly implement inside component?

**Comment by damoncro** [comment]
@leejw51crypto 
UWallet is necessary. It is a different type than UWalletComponent, making use of `State Pattern` to avoid calling wrong functions, e.g. users can not call `GetAddress` before creating the wallet, and it is more natural in blueprint, like below (notice the Red Square line):

image

**Comment by damoncro** [comment]
@leejw51crypto ok, add more docs about emphasizing `WalletComponent` is only for testing purpose, both in gitbook and source codes comments. 

This component design PR is for refactoring the original DefiWalletCoreActor based design, aka, simply replicate/split functions from the DefiWalletCoreActor into components based, and creating a more modular design, so most functions inDefiWalletCoreActor would directly apply to the component without adding breaking changes. 

To address other general issues, such as security issues, network issues, replication issues, performance issues etc, would be better to be addressed in another PR.

**Comment by damoncro** [comment]
In my understanding, Component is simply part of actor, but is more abstract, like other components in unreal, it represents the `common behaviors` of the actor, after attaching to it as a sub-object.
https://docs.unrealengine.com/4.27/en-US/ProgrammingAndScripting/ProgrammingWithCPP/UnrealArchitecture/Actors/Components/.

Users can attach component to any actor, and then duplicate that actor to any locations in the map.

There is no difference that users replicate actors - specifically, DefiWalletCoreActor, DefiWalletCoreActor can be spawned or duplicated in the map as many as users want.

So above problems exist in actors would still happen in component.

But component is behaviour based, from bottom to top (We have the behaviour first), while actor is from top to bottom (we have the `thing`, e.g. item, weapon, etc first), both designs are acceptable. But like I mentioned, if the users simply want to add blockchain functions to the any actor they want, instead of creating a sub-class from DefiWalletCoreActor in Tutorial (See https://docs.cronos.org/cronos-play/getting-started_unreal/quick-start/actors-and-blueprint-classes), they can choose the component and attach to the blockchain component into the actor they want to work with.

Correct me if I am wrong.

**Comment by damoncro** [comment]
The issue may be concerned is that users may forget which actor attach the blockchain component, and duplicate a lot of that actors, and boom... unknown behaviours, lots of network calls races, probably?

But I guess any actor has the issue? We also can not avoid the user from replicating DefiWalletCoreActor.

To address the issue, instead of avoiding the users from replicating actors by forcing users who can only use one global singleton actor, but address in api level, in network layer? Not found a good solution at this moment.

**Comment by damoncro** [comment]
`__func__`  is platform dependent, not tested.

File: Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp [hunk 1]
```diff
@@ -408,10 +408,6 @@ void ADefiWalletCoreActor::InitializeNewDevelopmentOnlyWallet(
     rust::cxxbridge1::String result =
         _coreWallet->get_address(CoinType::CryptoOrgMainnet, 0);
     output = UTF8_TO_TCHAR(result.c_str());
-    // zeroize mnemonics
-
-    zeroize_buffer((char *)result.data(), 0, result.size());
```
Comment (tomtau) [inline Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:413]: why not to zeroize when not in use?

Comment (damoncro) [reply Source/CronosPlayUnreal/Private/DefiWalletCoreActor.cpp:413]: It is misleading because the `result` here is the `address` instead of `mnemonics`.

**Comment by leejw51crypto** [comment]
component is unreal object behaviour, which can be added to many objects.

but i think wallet is too confidential to be a behaviour.

it should be blackbox which is hidden local only.

if it's used behaviour, it will be added to playercontroller.
but playercontroller is packaged in game-mod.

and it replicated to all networking participants

**Comment by leejw51crypto** [comment]
WallletComponent, which communicates with main wallet instance would be good?

**Pull request closed without merging**